### PR TITLE
config: add a Enabled option to LogConfig, run logmon optionally

### PR DESCRIPTION
**Pull request by towe75**
Closes https://github.com/hashicorp/nomad/issues/11175

### CODE REVIEW

File: website/content/docs/job-specification/logs.mdx [hunk 1]
```diff
@@ -43,6 +43,9 @@ please see the [`nomad alloc logs`][logs-command] command.
 
 ## `logs` Parameters
 
+- `enabled` `(bool: true)` - Controls if Nomad should write and rotate logfiles.
+  You can disable this setting if you're running e.g. some thirdparty log manager.
```
Comment (tgross) [inline website/content/docs/job-specification/logs.mdx:47]: Can we discuss a little bit what the envisioned workflow is? Let's assume we're using something like promtail... how would the operator configure this to capture the stdout/stderr from a task?

Comment (towe75) [reply website/content/docs/job-specification/logs.mdx:47]: Sure. The option is merely a resource optimization by not running a logmon process per task. The exact way how somebody can use e.g. promtail depends on the task driver. 

Docker and podman driver, for instance, allow you to direct the container log to the hosts journal.  Promtail can easily collect it from there. At least podman will annotate each line with a whole load of additional labels so that promtail/loki can provide appropriate log streams. Additional both drivers allow you to add your own key/value pairs to each journal entry etc. 
For docker you can AFAIK also rely on log plugins to do the heavy lifting.

There is also a different use case if you're running "system containers" instead of "application containers". A system container has it's own init process (e.g. systemd) and usually it's own internal log (e.g. journal). Such a container can run multiple services and redirecting the entire log as a "text stream" would throw away journal metadata. In this situation one might point promtail directly to the in-container journal folder (or  a log volume) to again scrape the data in great detail. 

All of that works without this PR - but it has the smell of a unnecessary process per container. It does not waste a lot of resources but it can surely be noticeable in high-density setups with many small containers on one host.

You might also want to have a look at the docker drivers "disable_log_collection" which has the same purpose. But it builds upon some exclusive internal driver interface that i can not use for the podman driver, hence the PR. Also the PR is more flexible because it offers a per-task solution.

I can elaborate more, if you want, but i think you understand my point now.

Comment (tgross) [reply website/content/docs/job-specification/logs.mdx:47]: Thanks, that helps. It'd probably be a good idea to extend this bit of the docs a little to something like:

```suggestion
- `enabled` `(bool: true)` - Controls if Nomad should write and rotate logfiles.
  You can disable this setting if you're running an external log manager (ex. `journald`)
  and the task can emit logs to that collector.
```

Comment (towe75) [reply website/content/docs/job-specification/logs.mdx:47]: Ok. What about the redundant disable_log_collection? Should we at least cross-link the two options?
Further: i did not test all task drivers but simply focused on podman. 

Also i am not sure about a possible UI integration. It would be nice to e.g. hide the "Logs" tab or to replace the log panel with some textual hint. Could also go into a new issue because it's a cosmetic change.

Comment (tgross) [reply website/content/docs/job-specification/logs.mdx:47]: > Ok. What about the redundant disable_log_collection? Should we at least cross-link the two options?

We'll end up with a docker logging shim running if we don't disable that as well. So what we'll probably want to do is:
* thread `logging.enabled = false` down into the driver configuration and have it set `disable_log_collection`
* add a note to the docs that the `disable_log_collection` field is deprecated and that you should use `logging.enabled = false`
* then we'll remove the `disable_log_collection` field in a future version of Nomad.

> Further: i did not test all task drivers but simply focused on podman.

Understood. I can check this branch out and do some smoke testing to verify there are no edge-case gotchas.

> Also i am not sure about a possible UI integration. It would be nice to e.g. hide the "Logs" tab or to replace the log panel with some textual hint. Could also go into a new issue because it's a cosmetic change.

I think it's probably ok to have that be a separate issue but I'm going to ping @ChaiWithJai on that for his opinion.

Comment (towe75) [reply website/content/docs/job-specification/logs.mdx:47]: I need a little bit of help with Nomads internal API. 

It's clear to me that i need to carry LogConfig.Enabled into the docker driver handle to have it available for RecoverTask() etc.  But i don't see how i could get a hold on LogConfig from within StartTask() without enhancing drivers.TaskHandle as well. 

Is there some tool method to get e.g. the alloc structure or the full job definition from within a driver plugin?

Comment (tgross) [reply website/content/docs/job-specification/logs.mdx:47]: It's a little gross. 😀 

The shared task driver [`TaskConfig`](https://github.com/hashicorp/nomad/blob/v1.1.6/plugins/drivers/driver.go#L265-L287) struct and associated [protobuffer](https://github.com/hashicorp/nomad/blob/v1.1.6/plugins/drivers/proto/driver.proto#L421-L479) are the data structures that contain the entire definition of the task that task drivers have to care about. But note that the task-driver-specific definition is in the `MsgpackDriverConfig` message which has been serialized to bytes and passed in the [`msgpack_driver_config`](https://github.com/hashicorp/nomad/blob/v1.1.6/plugins/drivers/proto/driver.proto#L430) field. This blob is mostly opaque to Nomad servers and clients, which is why there are all kinds of little annoying misfeatures around `task.config`.

So we'll want to pass that value in on the `TaskConfig` struct. It should default to enabled, which in the nil-is-unset world of protobufs mean the flag should be named "disabled" so that the default value is correct. Task drivers can then react to that value if they're doing anything special with logs (as the `docker` driver does).

We might want to consider passing a `LogConfig` message that includes the enabled/disabled flag, rather than just doing a bool flag on `TaskConfig`. That way we can easily add other log features to the task driver interface without having a bunch of top-level flags. So something like:

```diff
diff --git a/plugins/drivers/proto/driver.proto b/plugins/drivers/proto/driver.proto
index f315d3215..e10e18595 100644
--- a/plugins/drivers/proto/driver.proto
+++ b/plugins/drivers/proto/driver.proto
@@ -418,6 +418,11 @@ message DNSConfig {
     repeated string options = 3;
 }

+message LogConfig {
+    // disable all task logging helpers, if supported by the driver
+    bool disabled = 1;
+}
+
 message TaskConfig {

     // Id of the task, recommended to the globally unique, must be unique to the driver.
@@ -476,6 +481,9 @@ message TaskConfig {

     // DNSConfig is the configuration for task DNS resolvers and other options
     DNSConfig dns = 17;
+
+    // LogConfig is the configuration of task logging
+    LogConfig log_config = 18;
 }

```

**Comment by tgross** [comment]
Closing this in lieu of https://github.com/hashicorp/nomad/pull/16962, but because I lifted a good bit of @towe75's work there I've marked him as a co-author on that PR.
